ENH: Convert gaussian_kde logpdf to Cython

europeanplaice opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
<!--Please explain your changes.-->
In this PR, I converted gaussian_kde.logpdf to Cython. The code is based on `gaussian_kernel_estimate` so it may not be fully optimized. To take advantage of Cython, I avoided built-in functions, but for example, I added `logsumexp` manually and introduced nested loops. Feedback is welcome.

#### Additional information
<!--Any additional information you think is important.-->
You couldn't see a significant performance improvement on middle-sized datasets. However, as far as I know, this Cython code can run faster especially on large-scale datasets.
I am going to run more detailed benchmark tests.

tupui reviewed (commented):
Thanks @europeanplaice for the PR. To measure the performance, you can have a look at our benchmark system https://scipy.github.io/devdocs/dev/contributor/benchmarking.html#benchmarking-with-asv

europeanplaice commented:
@tupui Thank you for your advice but, currently, I can't build scipy in my environment. So I ran a benchmark test here.
https://colab.research.google.com/drive/1ZJwqlwW7n5DueNNaeOI1nWDf_4qt4_9K?usp=sharing

europeanplaice commented:
This is the benchmark test I ran with asv in my local environment. At least for my machine, we can see an improvement.


![compare](https://user-images.githubusercontent.com/38364983/152541961-97367e8b-b149-411e-9bb4-5d1f436fa4d4.png)

tupui commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,92 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    m = xi.shape[0]
+    p = values.shape[1]
+
+    if xi.shape[1] != d:
+        raise ValueError("points and xi must have same trailing dim")
+    if precision.shape[0] != d or precision.shape[1] != d:
+        raise ValueError("precision matrix must match data dims")
```
We usually keep input validation in Python.

europeanplaice commented:
Thank you @tupui. Please correct me if I am wrong.

https://github.com/scipy/scipy/blob/a2b53047a27897d6dae1587f4b88549934f759f7/scipy/stats/_kde.py#L590-L601
and
https://github.com/scipy/scipy/blob/a2b53047a27897d6dae1587f4b88549934f759f7/scipy/stats/_stats.pyx#L795-L796

is the same tests. So the test in cython can be removed.
In my commit, I removed these lines.

***
Below also could be moved to `logpdf` but, it is useless because it is clear that the covariance's shape corresponds to the dataset's shape and, I can't come up with the other case. So I think it also can be removed.
In my commit, I removed these lines.
https://github.com/scipy/scipy/blob/a2b53047a27897d6dae1587f4b88549934f759f7/scipy/stats/_stats.pyx#L797-L798

***
Below, I can't validate these lines raise an error because `np.common_type` returns
<class 'numpy.float16'> itemsize = 2 or
<class 'numpy.float32'> itemsize = 4 or 
<class 'numpy.float64'> itemsize = 8 or
<class 'numpy.float128'>itemsize = 16
and `self.covariance` must be float64 so I can't make invalid data that raise an error.
In my commit, I didn't touch it but didn't add tests.
https://github.com/scipy/scipy/blob/a2b53047a27897d6dae1587f4b88549934f759f7/scipy/stats/_kde.py#L603-L613

tupui commented on scipy/stats/_kde.py:
```diff
@@ -596,30 +596,25 @@ def logpdf(self, x):
             else:
-                msg = "points have dimension %s, dataset has dimension %s" % (d,
-                    self.d)
+                msg = f"points have dimension {d}, " \
+                      f"dataset has dimension {self.d}"
```
```suggestion
                msg = (f"points have dimension {d}, "
                       f"dataset has dimension {self.d}")
```

tupui commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,87 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+                estimate_[j, k, i] = log_values_[i, k] + log_arg
+
+                # Collect max value for logsumexp
+                if max_[j, k] < estimate_[j, k, i]:
```
So we keep updating `max_[j]` as we proceed on `i,k` right?

europeanplaice replied:
I will include this suggestion in the codes, thank you.

tupui commented on scipy/stats/tests/test_kdeoth.py:
```diff
@@ -347,6 +347,20 @@ def test_kde_output_dtype(point_type, dataset_type, weights_type, bw_type):
+    gkde = stats.gaussian_kde(xn)
+    xs = np.random.randn(3, 10)
+
+    # Act / Assert
+    with pytest.raises(ValueError,
+                       match="points have dimension 3, "
+                             "dataset has dimension 2"):
```
```suggestion
    msg = "points have dimension 3, dataset has dimension 2"
    with pytest.raises(ValueError, match=msg):
```

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,87 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    log_values_ = np.zeros((n, p), dtype)
+    for i in range(n):
+        for k in range(p):
+            log_values_[i, k] = math.log(values_[i, k])
```
Though I have not experienced well in Cython, when we compile codes with Cython, math looks faster than NumPy.
Here is the benchmark code in Colab. 
https://colab.research.google.com/drive/1r8V2bope5fWIpZAoOK7tny5d2_kzXelO?usp=sharing

tylerjereddy replied:
this is pretty common yeah--using the standard C math library for custom-written Cython loops, so should be "ok"

tupui replied:
Thanks Tyler 👍

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,87 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+        log_norm += math.log(whitening[i, i])
+
+    # Create the result array and evaluate the weighted sum
+    estimate = np.zeros((m, p), dtype)
+    max_ = np.zeros((m, p), dtype) - np.inf
+    estimate_ = np.zeros((m, p, n), dtype)
```
Thank you. I changed them to `empty` and checked but consequently some values of the result array became `nan`. I'm not sure why.

tupui replied:
It's because of the `- np.inf` I suppose.

p.s. please don't resolve the conversations. It's better to let reviewers close them so they can keep track of things they want to look at.

tupui replied:
Actually yes, `max` is ok to be zeros because of the comparison. But `estimate_` has no need to be.

europeanplaice replied:
Sorry for resolving the conversations. I reopened all of them. I'll leave it to you and reviewers to close them.

europeanplaice replied:
For `estimate_`, I changed `zeros` to `empty` at [ccf299f](https://github.com/scipy/scipy/pull/15493/commits/ccf299feeb54c8517aa47736084029762119c005). Thanks.

tupui replied:
No problem, all communities work differently with conversations 😅

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,87 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def gaussian_kernel_estimate_log(points, values, xi, precision, dtype, real _=0):
+    """
+    def gaussian_kernel_estimate_log(points, real[:, :] values, xi, precision)
+    Evaluate the log of the estimated pdf on a provided set of points.
+    Parameters
```
Did you mean this part `def gaussian_kernel_estimate_log(points, real[:, :] values, xi, precision)`? Actually I didn't care for it but I merely copied from `gaussian_kernel_estimate`. 

I think this line could be `def gaussian_kernel_estimate_log(points, values, xi, precision)`

For reference, you can refer to https://cython.readthedocs.io/en/stable/src/userguide/language_basics.html#python-functions-vs-c-functions .

europeanplaice replied:
After Cython code is compiled, some editors can't show arguments in the pop-up hint properly, but they would show like this.
`fused_cython_function: gaussian_kernel_estimate_log(*args, **kwargs)`

To show arguments, including a signature in a docstring is a better way.

In this discussion, a user showed the idea of including a signature in a docstring.
https://groups.google.com/g/pystatsmodels/c/dpVe6e0DaAM/m/FrInuMQFyB8J

The discussion was in 2013, and there may exist other solutions.

tupui replied:
Thanks for the reference. Since it's from Ralf (not just a user but our main maintainer I would say 😉) no problem. @tylerjereddy do we still need to do this though. The discussion is old so this might have changed. In the recent Cython I've seen been included we don't do this. But if it's a best practice, I am happy to consider this for next time.

europeanplaice replied:
Some functions contain a signature in the docstrings.
Like below
https://github.com/scipy/scipy/blob/9bb9189a710130f6b00b7dcde74a83d375e1e282/scipy/spatial/_ckdtree.pyx#L671-L677
https://github.com/scipy/scipy/blob/9bb9189a710130f6b00b7dcde74a83d375e1e282/scipy/spatial/_ckdtree.pyx#L850-L856

Others don't have it.
https://github.com/scipy/scipy/blob/9bb9189a710130f6b00b7dcde74a83d375e1e282/scipy/signal/_peak_finding_utils.pyx#L20-L30

It seems that there are no specific conventions authors follow.

tupui replied:
@europeanplaice exactly, that's why I am asking here if we should prefer one approach over the other. Although I have added Cython in the library too, I am not a power user and it would be good to know what is the recommended approach.

europeanplaice replied:
I'm also not a professional user, but in my opinion, it's ok to have a signature in docstrings if there is not a rule of it.
Although it looks a bit odd, I don't think having a signature in the docstrings reduces readability as help for users.
In my environment, the docstring is shown like this.
![image](https://user-images.githubusercontent.com/38364983/156000949-0ddf2fbd-fb33-4cc6-aab9-bada1d3a8bea.png)

rgommers replied:
The functions/methods that do have a signature in the docstring are the ones that are directly providing public API. That's when it is needed for the html docs. In this case, it's a private function that's used from pure Python code within SciPy, so there's no point in adding the signature.

rgommers replied:
Although, in this particular case, I think the point was to express the `real[:, :]` because that's now hidden by the fused types usage in an unusual way.

rgommers replied:
I'll fix the spacing and keep the signature.

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,87 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+        # Calculate logsumexp
+        for k in range(p):
+            for i in range(n):
+                estimate[j, k] += math.exp(estimate_[j, k, i] - max_[j, k])
```
Absolutely. Thank you for the clever fix!

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,86 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+
+        # Calculate logsumexp
+        for k in range(p):
+            for i in range(n):
+                estimate[j, k] += math.exp(estimate_[k, i] - max_[k])
+            estimate[j, k] = math.log(estimate[j, k])
```
Thank you for checking the code in details. However, I don't agree with you because 
`estimate[j, k] = math.log(estimate[j, k])` don't override the loop.
In the for loop, `math.exp(estimate_[k, i] - max_[k])` for each `i` is added to `estimate[j, k]`
The next line runs log transformation of this result.

If this line was `estimate[j, k] = math.log(estimate_[j, k])`, this would override the loop.

Correct me if I am wrong...

tupui replied:
Oh, yes I read this too fast. Sorry for the noise you're right.

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,87 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    values_ = values.astype(dtype, copy=False)
+
+    # Log transformation of values_. It is faster than np.log(values_).
+    log_values_ = np.zeros((n, p), dtype)
```
Thanks! I included your suggestion.

tupui reviewed (approved):
Thanks @europeanplaice. LGTM, but it would be good if someone else could have a look. I added a milestone to hopefully get more attention.

Note: the benchmark should be refactored with a param instead of having 2 different set of tests for low/high number of points. A bit out of scope but can be a followup if you are interested.

mdhaber commented on scipy/stats/_kde.py:
```diff
@@ -596,30 +596,27 @@ def logpdf(self, x):
-                result[i] = logsumexp(0.5 * log_to_sum)
-
-        return result
+            raise ValueError(
```
Can this be tested? Or, if it is already tested, can we include a `match` to make sure the intended message is emitted?

europeanplaice replied:
Partially, it can be tested. 
I copied it from `evaluate` method so that `logpdf` should be consistent with `evaluate` and I don't have any reason to write this code except for it.

`np.common_type` returns one of `[_nx.half, _nx.single, _nx.double, _nx.longdouble, _nx.csingle, _nx.cdouble, _nx.clongdouble]`. And its `itemsize` is also one of them
```python
import numpy.core.numeric as _nx
print(np.dtype(_nx.half), np.dtype(_nx.half).itemsize)
print(np.dtype(_nx.single), np.dtype(_nx.single).itemsize)
print(np.dtype(_nx.double), np.dtype(_nx.double).itemsize)
print(np.dtype(_nx.longdouble), np.dtype(_nx.longdouble).itemsize)
print(np.dtype(_nx.csingle), np.dtype(_nx.csingle).itemsize)
print(np.dtype(_nx.cdouble), np.dtype(_nx.cdouble).itemsize)
print(np.dtype(_nx.clongdouble), np.dtype(_nx.clongdouble).itemsize)
```
```
float16 2
float32 4
float64 8
float128 16
complex64 8
complex128 16
complex256 32
```
It means if np.dtype is complex256 or float16 it raises an error. 


For complex256 , the test would be like,
```python
def test_pdf_logpdf_dtype_validation():
    xn = (np.random.random(10) + np.random.random(10) * 1j).astype("clongdouble")
    xs = (np.random.random(10) + np.random.random(10) * 1j).astype("clongdouble")
    gkde = stats.gaussian_kde(xn)
    msg = "<class 'numpy.complex256'> has unexpected item size 32"
    with pytest.raises(ValueError, match=msg):
        gkde.logpdf(xs)
```

But, for float16 it can't be tested. `gkde.covariance` is float64 so np.common_type returns float64 even if an array's dtype is float16.

mdhaber commented on scipy/stats/_kde.py:
```diff
@@ -596,30 +596,27 @@ def logpdf(self, x):
-            result = logsumexp(0.5 * log_to_sum, axis=1)
+        output_dtype = np.common_type(self.covariance, points)
+        itemsize = np.dtype(output_dtype).itemsize
+        if itemsize == 4:
```
Are these already tested?

europeanplaice replied:
`test_kde_output_dtype` didn't address these lines because this test check the behavior of not `logpdf` (or `evaluate`) but `stats.gaussian_kde`.

rgommers replied:
This `itemsize` block looks like a very weird check, almost certainly not the right thing to do. What is it trying to achieve?

europeanplaice replied:
Thank you for the review. I copied this block from `evaluate(pdf)` so that this function is consistent with `pdf`. `logpdf` and `pdf` have the same functionality except for output values. In this sense, I thought the implementations of `logpdf` and `pdf` also should be similar to each other. I understand this block looks meaningless, and I agree to remove this block. However, to decide whether to remove it, we may have to check how the https://github.com/scipy/scipy/pull/10824 (I copied from this PR) was processed.

rgommers replied:
Makes sense, I looked at the history and it's okay to keep. I'll push a refactor to avoid having two copies of that code block floating around.

europeanplaice replied:
Thank you!

mdhaber reviewed (approved):
I can't review the algorithmic details or Cython, but the Python side looks OK. I'm not very familiar with the `gaussian_kde` code, so is it already well tested for `logpdf` accuracy? ~~If not, I'd add accuracy tests.~~ Yes it is tested against `pdf`. I will also check whether `kernel.logpdf(values)` for the example in the docstring changes much. _Update: maximum difference is `1.776e-15`, no problem there. Speedup is a factor of ~2x (35.3 ms vs 84.8 ms)_

@tupui if these are tested to your satisfaction, feel free to merge unless you'd like to wait for a more thorough review from someone qualified to go through the algorithm line by line.

tylerjereddy reviewed (commented):
Reviews seem mostly positive here--the duplicated `asv` benchmark is maybe a bit awkward?

europeanplaice commented on benchmarks/benchmarks/stats.py:
```diff
@@ -426,6 +426,14 @@ def time_gaussian_kde_evaluate_many_points(self):
+
+    def time_gaussian_kde_logpdf_many_points(self):
+        # test gaussian_kde logpdf on many points
+        self.kernel.logpdf(self.positions)
```
Thank you and I agree with you. I added `param` to parameterize the benchmarks.

tylerjereddy commented:
The doc build warnings don't look related--I believe we recently moved this stuff around somewhere else so that may go away with a rebase:

```
/home/circleci/repo/doc/source/tutorial/linalg.rst:354: WARNING: py:obj reference target not found: linalg.pinv2
/home/circleci/repo/doc/source/tutorial/linalg.rst:404: WARNING: py:obj reference target not found: linalg.pinv2
```

I approved the CI matrix to run again here

rgommers commented:
I'll have a look at getting this merged now.

rgommers commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,90 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    estimate = np.zeros((m, p), dtype)
+    zeros_max = np.zeros((p,), dtype) - np.inf
+    empty_estimate_ = np.empty((p, n), dtype)
+    for j in range(m):
```
The loop variables `i` and `j` are exactly reversed compared to `gaussian_kernel_estimate`, can you please make them the same?

europeanplaice replied:
If I corrected the order of `i` and `j`, it would not work well because
```python
# Calculate logsumexp
for k in range(p):
    for i in range(n):
        estimate[j, k] += math.exp(estimate_[k, i] - max_[k])
    estimate[j, k] = math.log(estimate[j, k])
    estimate[j, k] += max_[k]
```
this block depends on `j`. Currently, the loop logic is complicated so it's hard to make it align with `gaussian_kernel_estimate`. The code would be simple if we could separate `logsumexp` from this loop.

mdhaber replied:
I made them the same.

rgommers commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,90 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+                residual = (points_[i, k] - xi_[j, k])
+                arg += residual * residual
+
+            log_arg = -arg / 2 + log_norm
```
`log_arg` isn't needed, please use `arg = -arg / 2 + log_norm` in analogy with `gaussian_kernel_estimate`

europeanplaice replied:
Thanks. I fixed it on https://github.com/scipy/scipy/pull/15493/commits/371f14337a5aeb91321ec2e6a0868aa5554ebbfa .

rgommers commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,90 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+            for k in range(p):
+                estimate_[k, i] = log_values_[i, k] + log_arg
+
+                # Collect max value for logsumexp
```
I don't understand what is going on here to be honest - can you explain?

rgommers replied:
It also has to do with `estimate`, `estimate_` and `empty_estimate_` - the copying and multiple arrays used there seem unnecessary. Why can this not be the same as for `gaussian_kernel_estimate`?

europeanplaice replied:
> I don't understand what is going on here to be honest - can you explain?

When it returns
$$ \log(\sum^N_{i=1} \exp(x_i)) $$ 
I see an underflow error because the value of $ x $ is extremely small.
To avoid this problem the formula should be changed as 

$$ \begin{aligned}
\log(\sum^N_{i=1} \exp(x_i)) 
&= \log\{\exp(x_{max})\sum^N_{i=1} \exp(x_i - x_{max})\} \\
& = \log\{\sum^N_{i=1} \exp(x_i - x_{max})\}  + x_{max}
\end{aligned}
$$ 

By this conversion, the range of the target of $ \exp $ is $ [xmin−xmax, 0] $ so, the possibility of an underflow can be reduced.

This function exists as `logsumexp` in scipy. However, scipy's  `logsumexp` is implemented in Python, to keep the calculate fast, I wrote it by myself. In this code block you quoted, it tries to find $ x_{max} $ .


In the calculation of $ log $ world, multiplication changes to addition.
`gaussian_kernel_estimate`s'
```python
 for k in range(p):
    estimate[j, k] += values_[i, k] * arg
```
corresponds to 
`gaussian_kernel_estimate_log`s'
```python
for k in range(p):
    estimate_[k, i] = log_values_[i, k] + log_arg
```

___

> It also has to do with estimate, estimate_ and empty_estimate_ - the copying and multiple arrays used there seem unnecessary. Why can this not be the same as for gaussian_kernel_estimate?

For each `m`, I have to reset `estimate_`. If we didn't reset it, it would be possible that the previous iteration's max value would be selected.

I'm not sure my explanation has enough information. Does it make sense?

rgommers replied:
Thanks @europeanplaice, yes that makes sense. It looks like this would have been much easier if we had `logsumexp` implemented in C/Cython and exposed in `special.cython_special`, right?

europeanplaice replied:
Yes. It would be great not only for this PR, but also for future contributions!

jjerphan replied:
Hi, I am just passing by to look for any PR involving Cython in scipy.

We do have [a Cython implementation for `logsumexp` in scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/ef21ca1950f9570f2f6b1ce05d24111f78f9f612/sklearn/linear_model/_sag_fast.pyx.tp#L76-L95). I guess you can reuse it. 

On a side note, I am available for reviewing Cython PRs for scipy. So if needed, feel free to `@` me. :)

tupui replied:
Thank you @jjerphan that's good to know 😃 💪

tupui replied:
Do you mean a place where a person would propose some reviewing expertise? I would even wider this and say, someone proposes a certain expertise. Sometimes we have features on our wish list that we don't know much, or have deep technical questions that do not need software eng. expertise at all.

That would be awesome. We could do something similar to Publon and think about how to motivate people to engage and do this. This is definitely something that Scientific Python think about.

cc @jarrodmillman

jjerphan replied:
I've created [this thread on the Scientific Python Discourse](https://discuss.scientific-python.org/t/pooling-contributors-expertise-between-projects/381) to broaden the discussions regarding sharing skills across projects among the ecosystem.

rgommers reviewed (changes requested):
I pushed some changes that were needed. The bigger issue is in the Cython code however - the new code is mostly (80% - 90%) the same as the code from `gaussian_kernel_estimate`, with a few discrepancies that are not easy to understand. @europeanplaice it seems much better to me to either make this a single routine that deviates only when there's a `log` yes/no, or if that's not easily possible, then refactor the common parts out or template the routine (`.pyx.in`) so two copies of this routine come from the same source.

tylerjereddy commented:
I'm going to bump the milestone here if that is alright. It sounds like Ralf has carved out a path forward for this PR at least.

mdhaber commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,90 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    xi : array_like with shape (m, d)
+        Coordinates to evaluate the estimate at in ``d`` dimensions.
+    precision : array_like with shape (d, d)
+        Precision matrix for the Gaussian kernel.
```
@europeanplaice if I'm reading this correctly, `precision` is the `inv_cov` matrix passed in during the call to `logpdf`. gh-5047 was attempting to avoid the use of `inv_cov`. Do you have a moment to take a look at what that PR was trying to do and comment on whether that is compatible with what you're doing here? I was going to take over and complete that PR, but I didn't want to cause merge conflicts here.

europeanplaice replied:
Hi, @mdhaber. Do you mean not https://github.com/scipy/scipy/issues/5047 but https://github.com/scipy/scipy/pull/5087?

mdhaber replied:
Yes, thanks! Sorry about that.

mdhaber replied:
@europeanplaice I'd suggest finishing this PR without worrying about this extra enhancement, then we can do it as a follow-up. 
Is there anything you need from us to finish this, btw?

europeanplaice replied:
@mdhaber 
For me it's good to finish this PR at this time. I don't have anything to add to it. Thanks for your suggestion.

mdhaber replied:
Ok, but it looks like there are some comments above (https://github.com/scipy/scipy/pull/15493#pullrequestreview-988546456) that don't have responses. Could you reply to those so we can ask Ralf to take a final look?

mdhaber commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,90 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    p = values.shape[1]
+
+    # Rescale the data
+    whitening = np.linalg.cholesky(precision).astype(dtype, copy=False)
+    points_ = np.dot(points, whitening).astype(dtype, copy=False)
+    xi_ = np.dot(xi, whitening).astype(dtype, copy=False)
```
If we want `np.dot(x, whitening)` but have only the covariance matrix `A` instead of its inverse, `precision`, we can do:
```python3
JAJ = A[::-1, ::-1]
L2 = linalg.cholesky(JAJ)
JL2J = L2[::-1, ::-1]
res2 = linalg.solve_triangular(JL2J, x.T, lower=True).T
```
I expect this will more numerically favorable as it avoids the matrix inversion, and although all the indexing/transposition might mean that copies are needed before LAPACK is called, it should be a bit faster otherwise because: 
* an inversion is replaced with Cholesky decomposition
* a cholesky decomposition is replaced with a triangular solve
* a matrix-matrix product is eliminated

Example:
<details>

```python3
import numpy as np
from scipy import linalg

np.random.seed(0)

n = 10
x = np.random.rand(n, n)
A12 = np.random.rand(n, n)
A = A12 @ A12.T

P = linalg.inv(A)
Lp = linalg.cholesky(P)
res = np.dot(x, Lp)

JAJ = A[::-1, ::-1]
L2 = linalg.cholesky(JAJ)
JL2J = L2[::-1, ::-1]
res2 = linalg.solve_triangular(JL2J, x.T, lower=True).T

np.testing.assert_allclose(res, res2)
```

</details>

mdhaber replied:
gh-16692 implements this idea for `pdf` and both benchmarks execute in half the time as main.

mdhaber commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,90 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+
+    # Evaluate the normalisation
+    log_norm = (- d / 2) * math.log(2 * PI)
+    for i in range(d):
+        log_norm += math.log(whitening[i, i])
```
Can this be:
```suggestion
    log_norm += np.diag(whitening).sum()
```
But to follow https://github.com/scipy/scipy/pull/15493/files#r908725848, I suppose I need to figure out how to compute the diagonal elements of the inverse of a matrix from the cholesky decomposition of the matrix. (Could `cho_solve` an identity matrix, but perhaps there is a better way?)

mdhaber reviewed (commented):
Thanks @europeanplaice. I think this is close. Just a few thoughts from me.

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,89 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    d = points.shape[1]
+    m = xi.shape[0]
+    p = values.shape[1]
+
```
I removed the shape checks according to this review, but I didn't touch `gaussian_kernel_estimate`. 
https://github.com/scipy/scipy/pull/15493#discussion_r800714903

mdhaber replied:
It looks like I may end up being the one merging this, so I added it back to suit my preference. I value the two functions being comparable more than I do saving lines. If we think it's important to remove these lines, it can be done for both functions in a separate PR.

europeanplaice commented on scipy/stats/_stats.pyx:
```diff
@@ -753,3 +753,89 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+            estimate[j, k] = math.log(estimate[j, k])
+            estimate[j, k] += max_[k]
+
+    return estimate
```
This change is from the review https://github.com/scipy/scipy/pull/15493#discussion_r804004070 .
However, as you said, we may have to apply these changes to `gaussian_kernel_estimate `

mdhaber replied:
This can be done in a separate PR.

europeanplaice commented:
Though I haven't written a unit test, I made a cython version `logsumexp`. It can only accept an one dimensional array and doesn't have ability comparable to `special.logsumexp`. However, it runs as fast as the previous commit and the loop inside `gaussian_kernel_estimate_log` can be simpler than before.

jjerphan commented:
Note that this PR overlaps on some portions with https://github.com/scipy/scipy/pull/16758.

mdhaber commented on scipy/stats/_stats.pyx:
```diff
@@ -769,3 +769,86 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
+    for i in range(d):
+        log_norm += math.log(whitening[i, i])
+
+    # Create the result array and evaluate the weighted sum
+    estimate = np.full((m, p), fill_value=-np.inf, dtype=dtype)
+    for i in range(n):
+        for j in range(m):
+            arg = 0
+            for k in range(d):
+                residual = (points_[i, k] - xi_[j, k])
+                arg += residual * residual
+
+            arg = -arg / 2 + log_norm
+            for k in range(p):
+                estimate[j, k] = logsumexp(estimate[j, k],
+                                           arg + log_values_[i, k])
```
I made this line-by-line equivalent with `gaussian_kernel_estimate`. If this resulted in a performance loss, please open a new PR to change it. I'd suggest getting a Cython-accessible `logsumexp` function addeded to `scipy.special` first, so it doesn't need to be part of this PR, and please demonstrate the difference in performance.

mdhaber commented:
@europeanplaice @jjerphan I editing things so that I'd be comfortable merging this. Please check that I did not make any huge mistakes, since I am not very fluent in Cython. If CI comes back green, I'll plan to merge on Monday.

europeanplaice commented:
I didn't find any mistake in the commit and it looks good to me. The code became simpler than before. Thank you.

jjerphan commented:
I won't be able to do a review in the coming days. Hence just move on without me if you need to.

mdhaber commented:
I'm that case I'll invite former reviewers @rgommers and @tupui to take a final look if they're interested. (If not, that's ok; I'm comfortable enough to merge if there are no more comments.)

mdhaber commented on scipy/stats/_kde.py:
```diff
@@ -255,21 +255,10 @@ def evaluate(self, points):
         result = gaussian_kernel_estimate[spec](
             self.dataset.T, self.weights[:, None],
-            points.T, self.cho_cov, output_dtype)
+            points.T, self.inv_cov, output_dtype)
```
```suggestion
            points.T, self.cho_cov, output_dtype)
```

mdhaber commented:
Thanks @europeanplaice and reviewers! I'll add an entry to the release notes.
